### PR TITLE
Sub-fields should not accept `include_in_all` parameter

### PR DESCRIPTION
**Pull request by makeyang**
Throw an exception when specifying `include_in_all` on multi-fields

Closes https://github.com/elastic/elasticsearch/issues/21710

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/mapper/TypeParsers.java [hunk 1]
```diff
@@ -229,6 +229,9 @@ public static void parseField(FieldMapper.Builder builder, String name, Map<Stri
                 builder.indexOptions(nodeIndexOptionValue(propNode));
                 iterator.remove();
             } else if (propName.equals("include_in_all")) {
+                if(parserContext.isWithinMultiField()) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/mapper/TypeParsers.java]: Nit: usually we put a space between `if` and `(`

File: core/src/main/java/org/elasticsearch/index/mapper/TypeParsers.java [hunk 2]
```diff
@@ -229,6 +229,9 @@ public static void parseField(FieldMapper.Builder builder, String name, Map<Stri
...
+                    throw new MapperParsingException("sub-fields shouldn't contain property: include_in_all, and current node is:" + name);
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/index/mapper/TypeParsers.java]: Have a look at the example below for copy_to about how we usually write these field names. We wrap all the variable bits of exception messages in `[]` almost everywhere. Right or wrong it is our habit and we should stick to it unless we have a good reason not to.

**Review by nik9000 — CHANGES_REQUESTED** [review]
Can you add a test for this? I'd probably make an `org.elasticsearch.index.mapper.IncludeInAllMapperTests` and use `org.elasticsearch.index.mapper.CopyToMapperTests` as an example.

**Comment by makeyang** [comment]
@nik9000  I modified code according to your comments. I'll commit test later. but I have one question related to CopyToMapperTests: it seems in this test, this exception: 
MapperParsingException("copy_to in multi fields is not allowed. Found the copy_to in field [" + name + "] which is within a multi field.")
is not test?
if so, why? if it is not the truth, tell me some details. thanks

**Comment by nik9000** [comment]
> is not test?

You mean that test doesn't exist? I suspect we just forgot to make it. If you'd like to make it while you are there that'd be lovely.

**Comment by makeyang** [comment]
@nik9000  it turns out that it is tested but in another test case file:MultiFieldCopyToMapperTests. so I add this new test case file:MultiFieldIncludeInAllMapperTests
please help to review. thanks

**Comment by makeyang** [comment]
@nik9000  can you help to review this?

File: core/src/test/java/org/elasticsearch/index/mapper/MultiFieldIncludeInAllMapperTests.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.elasticsearch.index.MapperTestUtils;
+import org.elasticsearch.test.ESTestCase;
+
+import java.io.IOException;
+
+import static org.elasticsearch.common.xcontent.XContentFactory.jsonBuilder;
+import static org.hamcrest.core.IsEqual.equalTo;
+
+/**
+ * Created by makeyang on 2016/12/8.
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/index/mapper/MultiFieldIncludeInAllMapperTests.java]: Can you replace this with something descriptive of what it does or remove it?

File: core/src/test/java/org/elasticsearch/index/mapper/MultiFieldIncludeInAllMapperTests.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+ */
+public class MultiFieldIncludeInAllMapperTests extends ESTestCase {
+    public void testExceptionForCopyToInMultiFields() throws IOException {
+        XContentBuilder mapping = createMappinmgWithIncludeInAllInMultiField();
+
+        // first check that for newer versions we throw exception if copy_to is found withing multi field
+        MapperService mapperService = MapperTestUtils.newMapperService(createTempDir(), Settings.EMPTY);
+        try {
+            mapperService.parse("type", new CompressedXContent(mapping.string()), true);
+            fail("Parsing should throw an exception because the mapping contains a include_in_all in a multi field");
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/index/mapper/MultiFieldIncludeInAllMapperTests.java]: We tend to use `expectThrows` for this when we write new tests. It isn't required but it is a bit cleaner to read.

**Review by nik9000 — CHANGES_REQUESTED** [review]

**Comment by makeyang** [comment]
@nik9000 can you help to review this please?

**Comment by makeyang** [comment]
@nik9000  can you help to review this or ask someone else to help to review it? thanks very much.

**Review by nik9000 — APPROVED** [review]

**Comment by nik9000** [comment]
Looks right to me!

elasticmachine, please test this.

**Comment by nik9000** [comment]
Thanks @makeyang! I've merged to master and will backport to 5.x:

master: 2e1d152fc0b50633e832b8785d961325e480b54e + 40b80ae10460567144bbb0d8dcb606d4a10418ac
5.x: backporting

**Comment by makeyang** [comment]
@nik9000  thanks very much.

**Comment by mfeltscher** [comment]
Isn't this a breaking change? This broke our mappings while upgrading from 5.1.x to 5.2.x

**Comment by clintongormley** [comment]
@mfeltscher thanks for reporting and yes, this shouldn't have gone into 5.2.  i've opened https://github.com/elastic/elasticsearch/issues/23654

**Pull request merged**